core-bundle/src/Util/SimpleTokenParser.php

```diff
@@ -137,25 +140,36 @@ private function canUseExpressionLanguage(array $data): bool
 
     private function evaluateExpression(string $expression, array $data, bool $canUseExpressionLanguage): bool
     {
-        if ($canUseExpressionLanguage) {
-            try {
-                return $this->expressionLanguage->evaluate($expression, $data);
-            } catch (\Exception $e) {
-                throw new \InvalidArgumentException($e->getMessage(), 0, $e);
-            }
+        if (!$canUseExpressionLanguage) {
+            return $this->evaluateExpressionLegacy($expression, $data);
+        }
+
+        $unmatchedVariables = array_diff($this->getVariables($expression), array_keys($data));
+
+        // If the expression contains variables that aren't provided, log and exit gracefully
+        if (!empty($unmatchedVariables)) {
+            $this->logUnmatchedVariables(...$unmatchedVariables);
+
+            return false;
```

Comment:
If we always return `false` if the condition contains unknown variables wouldn’t that flip the logic of inverted comparisons like `{if foo!=""}`?

Reply:
Does this work right now?

Reply:
I think so because unknown values are treated as `NULL` I guess. But I didn’t test it yet.

Reply:
Instead of returning `false` we could set `$data[$unmatchedVariable] = null;` for every value of `$unmatchedVariables` and afterwards evaluate the expression. WDYT?

Reply:
If so, we could just preset all missing vars with `null` and continue (instead of returning), I guess.

Reply:
Changed in ec3d58dc3f6e2834c1ed7df3f9e363b9a5dcd907.

Reply:
It seems that we have to fix that in Contao 4.4/4.9 too: https://github.com/contao/contao/blob/2afe13cdc425cd4da774df205d6df4758f5e47ca/core-bundle/src/Resources/contao/library/Contao/StringUtil.php#L603

We introduced that bug in Contao 4.2.3: https://github.com/contao/core-bundle/commit/b68d28266bda83da2cf0cdf03df9d37c809b9464

Reply:
see https://github.com/contao/contao/issues/2402